README.md

```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
+    If `<name>` has less than 4 characters, append `-` to the `<name>` until it's 4 characters long and use that as `<4-prefix name>`
+- Additionally also move all paths transitively referenced by the default Nix file to `pkgs/unit/<4-prefix name>/<name>/???` [TODO](https://github.com/nixpkgs-architecture/simple-package-paths/issues/19)
+- Remove the definition of that attribute in `pkgs/top-level/all-packages.nix`
+- For each moved path, create a compatibility layer from the old to the new path, potentially using a symlink. See [compatibility] for more details
 
-will be become eligible to be transformed as follows:
-- Move `pkgs/some/dir` to `pkgs/auto/<name>`
-- Fix any references to that folder (e.g. in update scripts) 
-- Remove the original definition
+These attributes will newly be added to `pkgs` by automatically calling `pkgs.callPackage pkgs/unit/<4-prefix name>/<name>/pkg-fun.nix { }` on all entries in `pkgs/unit`. In order to make this more efficient, `builtins.readDir` should be optimized as described [here](https://github.com/NixOS/nix/issues/7314).
 
-These attributes will newly be defined by listing all directories in `pkgs/auto` using `builtins.readDir` and calling `pkgs.callPackage` on all of them, which then gets added to the `pkgs` scope.
+## Compatibility layer
+[compatibility]: #compatibility-layer
+
+TODO: Nix files should use `import` to act like a symlink while also giving a warning with `builtins.trace`. Something like
 
-Attributes whose definition aren't of the above form won't be changed, so e.g. the following definition in `all-packages.nix` won't be changed:
 ```nix
-{
-  syncplay-nogui = syncplay.override { enableGUI = false; };
-}
+builtins.trace "warning: Using deprecated path ${./.}, use pkgs/unit/<name> instead, this will be removed after NixOS 22.05"
+  (import ../../pkgs/unit/name)
 ```
 
-However, if such definitions can be refactored into the above form they will become eligible for the transformation.
-
-## Backwards compatibility symlinks
-[symlinks]: #backwards-compatibility-symlinks
 
-When moving `pkgs/some/dir/default.nix` to the new `pkgs/unit/<name>/pkg-fun.nix`, a symlink will be created pointing from `pkgs/unit/some/dir/default.nix` to `pkgs/unit/<name>/pkg-fun.nix`. Reasoning:
+When moving `pkgs/some/dir/default.nix` to the new `pkgs/unit/<4-prefix name>/<name>/pkg-fun.nix`, a symlink will be created pointing from the old to the new location. Reasoning:
 - Current community discussions referencing old files from the `master` branch are still valid for some time. While GitHub doesn't provide an easy way to navigate to a symlink, seeing the path to where the file has moved is better than getting an error.
 - It provides an opportunity for code referencing old paths to be updated. While it's not possible to give a deprecation warning with symlinks, users will at least be able to read it in the NixOS release notes. This doesn't occur often in practice.
 
```

Comment:
Code that does not expect a symlink due to bad writing will break. Probably acceptable.

How to make sure we do not collect dangling symlinks even in that one NixOS release? If anything gets moved afterwards I don't expect people to think about the symlink myself including.

Reply:
That's a good point, this increases the complexity of this compatibility layer by a lot. What we could do is limit it to exactly one release and remove it right after. This way we only need to check/correct for dead symlinks/Nix files once when the branch-off happens and can remove them in master afterwards. But yes this adds some complexity.

Reply:
Discussed in https://github.com/nixpkgs-architecture/simple-package-paths/pull/20#discussion_r1028999091